include/oneapi/dpl/internal/random_impl/philox_engine.h

```diff
@@ -0,0 +1,483 @@
+    generate_internal()
+    {
+        result_type __loc_result;
+        scalar_type __curr_idx = state_.idx;
+
+        for (int __elm_count = 0; __elm_count < _N; __elm_count++)
+        {
+            if (__curr_idx == word_count)
+            { // empty buffer
+                philox_kernel();
+                increase_counter_internal();
+                __curr_idx = 0;
+            }
+            __loc_result[__elm_count] = state_.Y[__curr_idx];
+            __curr_idx++;
+        }
+
+        state_.idx = __curr_idx;
+
+        return __loc_result;
+    }
+
+    /* generate_internal() specified for a scalar output */
+    template <unsigned int _N = 0>
+    ::std::enable_if_t<(_N == 0), result_type>
+    generate_internal()
+    {
+        result_type __loc_result;
+        scalar_type __curr_idx = state_.idx;
+
+        if (__curr_idx == word_count)
+        { // empty buffer
+            philox_kernel();
+            increase_counter_internal();
+            __curr_idx = 0;
+        }
+
+        // There are already generated numbers in the buffer
+        __loc_result = state_.Y[__curr_idx];
+        state_.idx = ++__curr_idx;
+
+        return __loc_result;
+    }
+
+    void
+    discard_internal(unsigned long long __z)
+    {
+        scalar_type __curr_idx = state_.idx % word_count;
+        unsigned long long __newridx = (__curr_idx + __z) % word_count;
+        if (__newridx == 0)
+        {
+            __newridx = word_count;
+        }
+
+        // check if we can't simply iterate the index in the buffer
+        if (__z >= word_count - state_.idx)
+        {
+            unsigned long long __counters_increment = __z / word_count;
+            __counters_increment += ((__z % word_count) + __curr_idx) / word_count;
+
+            if (state_.idx < word_count)
+            {
+                __counters_increment--;
+            }
+
+            increase_counter_internal(__counters_increment);
+
+            if (__newridx < word_count)
+            {
+                philox_kernel();
+                increase_counter_internal();
+            }
+        }
+
+        state_.idx = __newridx;
+    }
+
+    /* Internal generation Philox kernel */
+    void
+    philox_kernel()
+    {
+        if constexpr (word_count == 2)
+        {
+            scalar_type __R0 = (state_.X[0]) & in_mask;
+            scalar_type __L0 = (state_.X[1]) & in_mask;
+            scalar_type __K0 = (state_.K[0]) & in_mask;
+            for (::std::size_t __i = 0; __i < round_count; ++__i)
+            {
+                auto [__hi0, __lo0] = internal::experimental::mulhilo<scalar_type, word_size>(__R0, multipliers[0]);
+                __R0 = __hi0 ^ __K0 ^ __L0;
+                __L0 = __lo0;
+                __K0 = (__K0 + round_consts[0]) & in_mask;
+            }
+            state_.Y[0] = __R0;
+            state_.Y[1] = __L0;
+        }
+        else if constexpr (word_count == 4)
+        {
+            scalar_type __R0 = (state_.X[0]) & in_mask;
```

Comment:
when the constants will be changed this part can be represented in a more generic way through a loop for n/2:
![image](https://github.com/user-attachments/assets/a746fd68-f32d-477d-a350-6eaf0bca7e81)